nixos/knot: add support for XDP setups

mweinelt opened the pull request:
## Description of changes

The Express Data Path (XDP) is a way to circumvent the traditional Linux
networking stack and instead run an eBPF program on your NIC, that makes
the decision to provide Knot with certain packets. This is way faster
and more scalable but comes at the cost of reduced introspection.

Unfortunately the `knotc conf-check` command fails hard with missing
interfaces or IP addresses configured in `xdp.listen`, so we disable it
for now, once the `xdp` config section is set. We also promote the config
check condition to a proper option, so our conditions become public
documentation, and we allow users to deal with corner cases, that we have
not thought of yet.

We follow the pre-requisites documented in the Knot 3.3 manual, and set
up the required capabilities and allow the AF_XDP address family.

But on top of that, due to our strict hardening, we found two more
requirements, that were communicated upstream while debugging this.

- There is a requirement on AF_NETLINK, likely to query for and reconfigure
  the relevant network interfaces
- Running eBPF programs requires access to the `bpf` syscall, which we
  deny through the `~@privileged` configuration.

In summary We now conditionally loosen the hardening of the unit once we
detect that an XDP configuration is wanted. And since we cannot
introspect arbitrary files from the `settingsFiles` option, we expose XDP
support through the `enableXDP` toggle option on the module.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mweinelt committed:
nixos/knot: add support for XDP setups

The Express Data Path (XDP) is a way to circumvent the traditional Linux
networking stack and instead run an eBPF program on your NIC, that makes
the decision to provide Knot with certain packets. This is way faster
and more scalable but comes at the cost of reduced introspection.

Unfortunately the `knotc conf-check` command fails hard with missing
interfaces or IP addresses configured in `xdp.listen`, so we disable it
for now, once the `xdp` config section is set. We also promote the config
check condition to a proper option, so our conditions become public
documentation, and we allow users to deal with corner cases, that we have
not thought of yet.

We follow the pre-requisites documented in the Knot 3.3 manual, and set
up the required capabilities and allow the AF_XDP address family.

But on top of that, due to our strict hardening, we found two more
requirements, that were communicated upstream while debugging this.

- There is a requirement on AF_NETLINK, likely to query for and configure
  the relevant network interface
- Running eBPF programs requires access to the `bpf` syscall, which we
  deny through the `~@privileged` configuration.

In summary We now conditionally loosen the hardening of the unit once we
detect that an XDP configuration is wanted. And since we cannot
introspect arbitrary files from the `settingsFiles` option, we expose XDP
support through the `enableXDP` toggle option on the module.

mweinelt committed:
nixos/tests/knot: test the XDP interface

We reconfigure the secondary nameserver VM to do all the same things that
it did before, but now over the XDP interface.

mweinelt commented:
@ofborg test knot

vcunat commented:
I wonder if we should introduce an explicit option that turns off configuration checks, and this option would default to `cfg.keyFiles == [] && !withXDP` currently.

mweinelt commented:
I think both situations break really hard under the configured conditions, so I can't image why we'd want to allow someone to enable them. Is it, because it gives people more chances to experiment with knot? Is it because then we expose the basis for the decision better?

vcunat commented:
I meant it for the ability to disable the checks.  There might be some more corner cases that we haven't discovered yet.

mweinelt commented:
Added two options in the latest change:

- `enableXDP` with the same default as before, but now everyone may enable it at their own discrection
- `checkConfig` as discussed

vcunat commented:
I like the approach.  It's nice to utilize NixOS's knowledge of the daemon configs in this way.

vcunat commented:
> run an eBPF program on your NIC

Nit: I believe almost no hardware supports doing that.  Certainly none of the cards that we've tested (mainly Intel cards).  

I see nothing wrong in the diffs.

mweinelt commented:
Right, that may be something cards like Bluefield or Netronome can do. Let me reword that.

vcunat commented:
Yes, I do remember from years ago that some Netronome card could do it and maybe it was the only one at that point.

SuperSandro2000 commented on nixos/modules/services/networking/knot.nix:
```diff
@@ -142,7 +141,35 @@ let
-      enable = mkEnableOption (lib.mdDoc "Knot authoritative-only DNS server");
+      enable = mkEnableOption "Knot authoritative-only DNS server";
+
+      enableXDP = mkOption {
```
Shouldn't this be readOnly since setting it to true manually does nothing?

mweinelt replied:
As long as we allow `settingsFIle` this option can be used to relax the system hardening as required for XDP to be set up.

mweinelt committed:
nixos/knot: refactor

- Stop using `with lib`
- Drop `lib.mdDoc`
- Use `escaepSystemdExecArgs` for escaping

SuperSandro2000 reviewed (approved):
I've tested this successful in a hetzner cloud VM but needed to roll it back because I'm using DDNS/DynDNS and didn't read the docs deep enough.

mweinelt commented:
Added a link to the limitations on the `enableXDP` option.

mweinelt commented on nixos/modules/services/networking/knot.nix:
```diff
@@ -142,20 +169,53 @@ let
+          Extends the systemd unit with permissions to allow for the use of
+          the eXpress Data Path (XDP).
+
+          ::: {.note}
+            Make sure to read up on functional [limitations](https://www.knot-dns.cz/docs/3.3/html/operation.html#id21)
+            when running in XDP mode.
+          :::
```
The link pins the 3.3 release, since the anchor jumps between release series.

vcunat replied:
I think we'll improve that upstream.

vcunat replied:
So we switch to this one?
https://www.knot-dns.cz/docs/3.3/singlehtml/index.html#mode-xdp-limitations
or
https://www.knot-dns.cz/docs/latest/singlehtml/index.html#mode-xdp-limitations
or perhaps extract the right major version?  :-D  (I don't have time right now)

mweinelt replied:
Did the latter.

vcunat commented on nixos/modules/services/networking/knot.nix:
```diff
@@ -210,19 +270,35 @@ in {
+          "CAP_SYS_ADMIN"
+          "CAP_IPC_LOCK"
+        ] ++ lib.optionals (lib.versionOlder config.boot.kernelPackages.kernel.version "5.11") [
+          "CAP_SYS_RESOURCE"
```
This could theoretically be tricky for people who do `nixos-rebuild switch` when kernel branch changes (instead of `nixos-rebuild boot`).  But it seems OK not to safeguard against that here, as it's quite a rare combination in this case: knot+XDP + _downgrading_ across 5.11 boundary.
